### PR TITLE
chores: fix various corner cases, increment test coverage, update dependencies

### PR DESCRIPTION
**Pull request by menduz**
- fix various corner cases
- increment test coverage
- update dependencies
- tested and fixed logic for status checks
- add strict types
- add support for reverse-proxies URL (X-Forwarded-For)
- add support to create reverse proxies returning the response of a fetch call

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "typescript.tsdk": "node_modules/typescript/lib"
+}
```
Comment (jmoguilevsky) [inline .vscode/settings.json:3]:
```suggestion
}

```

File: src/logic.ts [hunk 2]
```diff
@@ -57,15 +59,19 @@ export function success(data: fetch.Response, res: ExpressModule.Response) {
...
     res.send(body)
   } else if (isBlob(body)) {
-    const blob = body as Blob
-    const stream = blob.stream()
-    if (stream.pipeTo) {
-      stream.pipeTo(res as any)
-    } else {
-      ;(blob.stream() as any).pipe(res)
-    }
+    // const blob = body as Blob
```
Comment (jmoguilevsky) [inline src/logic.ts:62]: why is this commented out?

File: src/logic.ts [hunk 3]
```diff
@@ -99,15 +105,15 @@ export const getRequestFromNodeMessage = <T extends http.IncomingMessage>(
 
   const requestInit: fetch.RequestInit = {
     headers: headers,
-    method: request.method.toUpperCase(),
+    method: request.method!.toUpperCase(),
```
Comment (jmoguilevsky) [inline src/logic.ts:108]: do we want to fail when no method is present? why not fallback to `GET`?
as in ```request.method?.toUpperCase() || 'GET'```

File: src/middleware.ts [hunk 4]
```diff
@@ -31,25 +15,22 @@ export function compose<Ctx>(...middlewares: Middleware<Ctx>[]): Middleware<Ctx>
...
 
-      if (!fn) return Promise.resolve({} as http.IResponse)
+      if (!fn) return {} as http.IResponse
 
-      try {
-        return Promise.resolve(fn(context, dispatch.bind(null, i + 1)))
-      } catch (err) {
-        return Promise.reject(err)
-      }
+      return await fn(context, dispatch.bind(null, i + 1))
```
Comment (jmoguilevsky) [inline src/middleware.ts:33]:
```suggestion
      return fn(context, dispatch.bind(null, i + 1))
```

File: src/router.ts [hunk 5]
```diff
@@ -160,7 +156,7 @@ export class Router<Context extends {}> implements IHttpServerComponent.MethodHa
     let router = this
 
     const hasPath = typeof middleware[0] === "string"
-    if (hasPath) path = (middleware.shift() as any) as string
+    if (hasPath) path = middleware.shift() as any as string
```
Comment (jmoguilevsky) [inline src/router.ts:159]: ah, typescript <3

File: src/status-checks.ts [hunk 6]
```diff
@@ -10,13 +38,57 @@ import { Router } from "./router"
...
+
+    const probes: { name: string; promise: Promise<boolean> }[] = []
+
+    let functionName: "startupProbe" | "readynessProbe" = startup ? "startupProbe" : "readynessProbe"
+
+    for (let c in components) {
+      if (typeof components[c][functionName] == "function") {
+        probes.push({
+          name: c,
+          promise: new Promise((ok) => {
```
Comment (jmoguilevsky) [inline src/status-checks.ts:67]: why not?

promise: async () => await components[c][functionName]!()

File: src/status-checks.ts [hunk 7]
```diff
@@ -10,13 +38,57 @@ import { Router } from "./router"
...
+          }),
+        })
+      }
+    }
+
+    const results = await Promise.all(probes.map(($) => $.promise))
+
+    const content: StandardStatusResponse = {
+      details: {},
+      status: results.some(($) => $ == false) ? "fail" : "pass",
```
Comment (jmoguilevsky) [inline src/status-checks.ts:80]:
```suggestion
      status: results.some(($) => $ === false) ? "fail" : "pass",
```

In this case you want to make sure it's `false` and not any falsy value

File: src/status-checks.ts [hunk 8]
```diff
@@ -10,13 +38,57 @@ import { Router } from "./router"
...
+    }
+
+    for (let it of probes) {
```
Comment (jmoguilevsky) [inline src/status-checks.ts:83]: TODO (can be out of scop of this PR): Paralelize this with a Promise all and a map of `await it.promise`

File: test/integration.spec.ts [hunk 9]
```diff
@@ -1,16 +1,201 @@
...
+    server.resetMiddlewares()
+    server.use(async (_, next) => {
+      next()
+      return next()
+    })
+    const res = await fetch.fetch(`/`)
+    expect(res.status).toEqual(500)
+  })
+
+  it("calling multiple next if there is no next returns 404", async () => {
```
Comment (jmoguilevsky) [inline test/integration.spec.ts:94]: why is this a 404?

File: src/logic.ts (hunk 2 above)
Comment (menduz) [reply src/logic.ts:62]: Because there is no test and I couldn't find a good use case for blobs in the server, I think it was meant for the browsers only. In any case, I prefered to fail explicitly

File: src/logic.ts (hunk 3 above)
Comment (menduz) [reply src/logic.ts:108]: I thought the same, here is my reasoning: method is only missing when `request: node.IncomingMessage` is not produced by http or https servers because it is a reused type. (the only other option I know is by using the `request` package).  Falling back to GET hides an underlying configuration issue of the server and would assume "GET" as default method. I opted to have the code failing here.

File: test/integration.spec.ts (hunk 9 above)
Comment (menduz) [reply test/integration.spec.ts:94]: imagine this scenario

```ts
routerV1.get('/v1', someHandler)
routerV2.get('/v2', someHandler)

server.use(authMiddleware)
server.use(rateLimitMiddleware)
server.use(routerV1.middleware())
server.use(routerV2.middleware())
```

What would you expect to happen if we call `GET /v1` `GET /v2` and `GET /v3`?

I expect `GET /v1 == 200`, `GET /v2 == 200` and `GET /v3 == 404`

Does it make sense?

File: src/status-checks.ts (hunk 8 above)
Comment (menduz) [reply src/status-checks.ts:83]: like line 76?
